### PR TITLE
fix(vpp-chaperone): delete-old-QoS-policers-if-policer-config-is-removed-from-node-configuration

### PR DESCRIPTION
**Pull request by Lahari-Kamasetty**
fix(vpp-chaperone): delete-old-QoS-policers-if-policer-config-is-removed-from-node-configuration
Signed-off-by: Lahari-Kamasetty <k.lahari@capgemini.com>

## Prerequisites

- [ ] I have read the [Contributing Guidelines](../blob/main/CONTRIBUTING.md).
- [ ] I have read the [Code of Conduct](../blob/main/CODE_OF_CONDUCT.md).
- [ ] If this is a non-trivial change, I have already opened an accompanying Issue.
- [ ] If applicable, I have included documentation updates alongside my code changes.

## Description

Delete old QoS policers if policer config is removed from node configuration

Modified the files : 
src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp
src/vpp_plugins/vpp-chaperone/VppConfigManager.h

## Test Plan

Verified on PUMA, PFA output file

[vppctl-show-policer-output.txt](https://github.com/terragraph/meta-terragraph/files/10175364/vppctl-show-policer-output.txt)

### CODE REVIEW

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 1]
```diff
@@ -792,6 +798,56 @@ void VppConfigManager::doCpeConfig (VppClient &vppClient)
               doCpePolicerConfig (vppClient, kv.first.asString (),
                                   kv.second["policers"]);
             }
+          else
+            {
+              vppClient.getClassifierTableIds (tableIds);
```
Comment (deltice) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:803]: The code itself looks good. Can you put this in a separate function for the sake of modularity/cleanliness? e.g. `deleteExistingPolicers` or something similar

**Review by deltice — CHANGES_REQUESTED** [review]

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 1 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comments

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h [hunk 2]
```diff
@@ -46,6 +46,10 @@ class VppConfigManager final
   void doCpePolicerConfig (VppClient &vppClient, const std::string &interface,
                            const folly::dynamic &policers);
 
+  // Delete Existing CPE Poliers.
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.h:49]: please fix typo "Poliers"

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h [hunk 3]
```diff
@@ -46,6 +46,10 @@ class VppConfigManager final
...
+  void deleteExistingPolicers (VppClient &vppClient,
+                               const std::string &cpe_interface);
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.h:51]: please follow naming format (camel case): `cpeInterface`

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h (hunk 2 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.h]: Addressed Review Comments

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h (hunk 3 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.h]: Addressed Review Comments

**Comment by deltice** [comment]
After looking at these pieces of code again, we identified another bug that should be fixed along with the initial issue. The `VppConfigManager::doCpePolicerConfig()` function is deleting all classifier tables each time it is called, not just classifier tables for the interface that gets passed in. The functionality just happens to be fine so far because we are only using one classifier table anyway so every time it is recreated, it has the same index as previously used by `setClassifierTableNetif` for other interfaces.

The fix for this new issue would be to use the "classify_table_by_interface" VPP API call instead of "classify_table_ids". This would result in creating a new classifier table for each interface. It makes sense to fix this newly identified issue with the original. It would also be best if functionality for deleting existing policers is addressed with changes inside `doCpePolicerConfig` rather than a separate function outside, and this would require some changes in `doCpeConfig` to call it a bit differently.

I can provide some more code pointers/assistance next week.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 4]
```diff
@@ -1644,89 +1651,90 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
   // otherwise VPP will crash if policed traffic is being passed while the
   // config is deleted.
   bool ifaceStopped = false;
-  // Delete old tables
-  std::vector<u32> tableIds;
-  vppClient.getClassifierTableIds (tableIds);
+  u32 swIfIndex = vppClient.ifaceToVppIndex (interface);
+  u32 tableId;
+  vppClient.getClassifierByIndex (swIfIndex, tableId);
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1656]: Check the return value. What does this return if the interface does not have a classifier table? Will the code below crash?

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h [hunk 5]
```diff
@@ -44,7 +44,7 @@ class VppConfigManager final
 
   // Add and delete CPE Policers.
   void doCpePolicerConfig (VppClient &vppClient, const std::string &interface,
-                           const folly::dynamic &policers);
+                           const folly::dynamic &policers, bool policerPresent);
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.h:47]: A better parameter name would be `isAdd` instead of `policerPresent` which is ambiguous

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 6]
```diff
@@ -1746,23 +1754,38 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
...
+          // be marked with DSCP corresponding to a traffic class with low drop
+          // preference (green)
+          VppClient::ClassifierSessionConfig sessionConfig;
+          std::vector<u8> matchVector = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+              0, 0, kAFxyMatch.at (tc), 0x80};
+          sessionConfig.match = matchVector.data ();
+          sessionConfig.hitNextIndex = policerIndex;
+          sessionConfig.tableIndex = tableConfig.tableIndex;
+          vppClient.addDelClassifierSession (sessionConfig);
+        }
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1780]: This if-elseif-if chain doesn't make sense. Write it like this:
```cpp
      if (policerPresent)
        {
          policerConfig.isAdd = 1; // Add
          vppClient.addDelPolicer (policerConfig, policerIndex);

          // Create session matching to AFx1: upstream traffic is expected to
          // be marked with DSCP corresponding to a traffic class with low drop
          // preference (green)
          VppClient::ClassifierSessionConfig sessionConfig;
          std::vector<u8> matchVector = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
              0, 0, kAFxyMatch.at (tc), 0x80};
          sessionConfig.match = matchVector.data ();
          sessionConfig.hitNextIndex = policerIndex;
          sessionConfig.tableIndex = tableConfig.tableIndex;
          vppClient.addDelClassifierSession (sessionConfig);
        }
      else
        {
          policerConfig.isAdd = 0; // Delete
          vppClient.addDelPolicer (policerConfig, policerIndex);
        }
```

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 7]
```diff
@@ -1625,12 +1630,14 @@ void VppConfigManager::setTctblEntry (VppClient &vppClient,
...
                                            const std::string &interface,
-                                           const folly::dynamic &policers)
+                                           const folly::dynamic &policers,
+                                           bool policerPresent)
 
 {
   u32 interfaceIndex = (u32)~0;
   auto ifaceMap = vppClient.getIfaceToVppIndexMap ();
   auto ifaceMapIter = ifaceMap.find (interface);
+  VppClient::PolicerConfig policerConfig;
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1640]: Move this back into the for loop - why did you move it to the top?

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 8]
```diff
@@ -790,7 +790,12 @@ void VppConfigManager::doCpeConfig (VppClient &vppClient)
             {
               LOG (INFO) << "Adding policers to CPE interface " << kv.first;
               doCpePolicerConfig (vppClient, kv.first.asString (),
-                                  kv.second["policers"]);
+                                  kv.second["policers"], true);
+            }
+          else if (!kv.second.count ("policers"))
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:795]: You don't need to check the same condition twice. Leave it as `else`.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 9]
```diff
@@ -1644,89 +1651,90 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
...
-        {
-          ifaceStopped = true;
-          VLOG (1) << "Stopping interface "
-                   << interface << " before removing old policer config";
-          vppClient.setInterfaceFlags (interface, false);
-        }
-      tableConfig.tableIndex = tableId;
-      tableConfig.isAdd = 0; // Delete old table
-      vppClient.addDelClassifierTable (tableConfig);
+      ifaceStopped = true;
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1661]: This check is always going to trigger once and no longer makes sense. You can probably remove it. However if `policerConfig == false` and there is no classifier already on this interface, we should probably return (is this right @deltice ?)

**Review by elludraon — CHANGES_REQUESTED** [review]

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 9 above)
Comment (deltice) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1661]: It can be okay to let the function continue if we care about cleaning up cases where classifier tables or policers might be configured using vppctl in some unexpected way.

If only using vpp_chaperone, it shouldn't be possible, for example, to have policers present while the classifier table is gone. It might be possible if someone uses vppctl to do things, I don't recall the exact combination that might lead to this though or what the behavior would be (maybe that policers exist and get shown by `vppctl show policer` but won't affect anything, but that could be confusing).

We still need to check if there is no classifier on the interface in either case. (To either return, or not stop the interface until we find a table/policer that does need to be deleleted.)

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.h (hunk 5 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.h]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 6 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 7 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 8 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 9 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 10]
```diff
@@ -1644,89 +1650,98 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
...
-  // Create table
-  tableConfig.isAdd = 1; // Add
-  // Skip 14 bytes for Ethernet header, 4 bits for IPv6 version header;
-  // match on full DSCP value contained in next 6 bits
-  std::vector<u8> maskVector = {0, 0, 0, 0, 0, 0, 0,    0,
-                                0, 0, 0, 0, 0, 0, 0x0F, 0xC0};
-  tableConfig.mask = maskVector.data ();
-  vppClient.addDelClassifierTable (tableConfig);
+  if (isAdd)
+    {
```
Comment (deltice) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1677]: Delete extra empty line

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 11]
```diff
@@ -1745,24 +1760,36 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
           VLOG (1) << "Remove old policer " << policerConfig.name;
           policerConfig.isAdd = 0; // Delete
           vppClient.addDelPolicer (policerConfig, policerIndex);
+
+          if (!isAdd)
+            {
+              continue;
+            }
```
Comment (deltice) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1767]: This looks unnecessary. The only remaining code inside this loop iteration is already inside an `if (isAdd)`

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 12]
```diff
@@ -1644,89 +1650,98 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
...
           vppClient.setInterfaceFlags (interface, false);
         }
       tableConfig.tableIndex = tableId;
       tableConfig.isAdd = 0; // Delete old table
       vppClient.addDelClassifierTable (tableConfig);
     }
+  else
+    {
+      LOG (ERROR) << "Interface does not have classifier table";
+    }
```
Comment (deltice) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1673]: This should not use `LOG (ERROR)` since it's not an error if an interface does not have a classifier table in this case.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 10 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

**Review by deltice — APPROVED** [review]

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 13]
```diff
@@ -790,7 +790,12 @@ void VppConfigManager::doCpeConfig (VppClient &vppClient)
             {
               LOG (INFO) << "Adding policers to CPE interface " << kv.first;
               doCpePolicerConfig (vppClient, kv.first.asString (),
-                                  kv.second["policers"]);
+                                  kv.second["policers"], true);
+            }
+          else
+            {
+              LOG (INFO) << "Deleting policers for CPE interface " << kv.first;
+              doCpePolicerConfig (vppClient, kv.first.asString (), 0, false);
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:798]: please pass `nullptr` instead of 0

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp [hunk 14]
```diff
@@ -1644,89 +1650,96 @@ void VppConfigManager::doCpePolicerConfig (VppClient &vppClient,
...
-  // Skip 14 bytes for Ethernet header, 4 bits for IPv6 version header;
-  // match on full DSCP value contained in next 6 bits
-  std::vector<u8> maskVector = {0, 0, 0, 0, 0, 0, 0,    0,
-                                0, 0, 0, 0, 0, 0, 0x0F, 0xC0};
-  tableConfig.mask = maskVector.data ();
-  vppClient.addDelClassifierTable (tableConfig);
-
+  else
+    {
+      LOG (INFO) << "Interface does not have classifier table";
```
Comment (elludraon) [inline src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp:1672]: Use `VLOG (1)` like above, otherwise this debug log cannot be disabled

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 13 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

File: src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp (hunk 14 above)
Comment (Lahari-Kamasetty) [reply src/vpp_plugins/vpp-chaperone/VppConfigManager.cpp]: Addressed Review Comment.

**Review by elludraon — APPROVED** [review]
Tested on Puma, this works now.
